lib/backend/atomicwrite.go

```diff
@@ -0,0 +1,245 @@
+func NotExists() Condition {
+	return Condition{
+		Kind: KindNotExists,
+	}
+}
+
+// Revision builds a condition that asserts the target key has the specified revision.
+func Revision(r string) Condition {
+	return Condition{
+		Kind:     KindRevision,
+		Revision: r,
+	}
+}
+
+// ActionKind marks the kind of an action to be taken.
+type ActionKind int
+
+const (
+	// KindNop indicates that no action should be taken.
+	KindNop ActionKind = 1 + iota
+
+	// KindPut indicates that the associated item should be written to the target key.
+	KindPut
+
+	// KindDelete indicates that any item at the target key should be removed.
+	KindDelete
+)
+
+// Action specifies an action to be taken against a backend item.
+type Action struct {
+	// Kind is the kind of action represented.
+	Kind ActionKind
+
+	// Item is the item to be written (only used when Kind is Put).
+	Item Item
+}
+
+// IsZero checks if this action appears unspecified.
+func (a *Action) IsZero() bool {
+	return a.Kind == 0
+}
+
+// Nop builds an action that does nothing.
+func Nop() Action {
+	return Action{
+		Kind: KindNop,
+	}
+}
+
+// Put builds an action that writes the provided item to the target key.
+func Put(item Item) Action {
+	return Action{
+		Kind: KindPut,
+		Item: item,
+	}
+}
+
+// Delete builds an action that removes the target key.
+func Delete() Action {
+	return Action{
+		Kind: KindDelete,
+	}
+}
+
+// ConditionalAction specifies a condition and an action associated with a given key. The condition
+// must hold for the action to be taken.
+type ConditionalAction struct {
+	// Key is the key against which the associated condition and action are to
+	// be applied.
+	Key []byte
+
+	// Condition must be one of Exists|NotExists|Revision(<revision>)|Whatever
+	Condition Condition
+
+	// Action must be one of Put(<item>)|Delete|Nop
+	Action Action
+}
+
+// Check validates the basic correctness of the conditional action.
+func (c *ConditionalAction) Check() error {
+	if len(c.Key) == 0 {
+		return trace.BadParameter("conditional action missing required parameter 'key'")
+	}
+
+	if c.Condition.IsZero() {
+		return trace.BadParameter("conditional action for %q missing required parameter 'condition'", c.Key)
+	}
+
+	if c.Action.IsZero() {
+		return trace.BadParameter("conditional action for %q missing required parameter 'action'", c.Key)
+	}
+
+	switch c.Condition.Kind {
+	case KindWhatever, KindExists, KindNotExists, KindRevision:
+	default:
+		return trace.BadParameter("conditional action for %q contains unexpected condition kind %v", c.Key, c.Condition.Kind)
+	}
+
+	switch c.Action.Kind {
+	case KindNop, KindDelete:
+	case KindPut:
+		if len(c.Action.Item.Value) == 0 {
+			return trace.BadParameter("conditional action for %q missing required put parameter Item.Value", c.Key)
+		}
+	default:
+		return trace.BadParameter("conditional action for %q contains unexpected action kind %v", c.Key, c.Action.Kind)
+	}
```

Comment:
Couldn't we move these checks up near the definition of `Condition` and `Action`? The `Check` method of some unrelated (kinda) type seems like a weird place to have to remember to update when adding a new variant of an enum.

Reply:
I chose to do this at this level because it made it easy to craft error messages that include the key.  We check the key first, and then all subsequent error messages can use the key to (hopefully) uniquely identify which specific conditional action actually failed the checks.  I think it improves ux.

Reply:
That's what `trace.Wrap` is for tho:
```go
if err := c.Condition.Check(); err != nil {
	return trace.Wrap(err, "checking condition for action on %q", c.Key)
}
```